### ISSUE TITLE
Floating-point parsing support in the std/parsing library

### ISSUE THREAD

**Issue by maximecb** [zetavm/zetavm#96]
A little while back, I spun off some of the lower-level parsing functions in the Plush parser into their own [parsing library](https://github.com/zetavm/zetavm/blob/master/plush/parsing.pls), so that other languages could reuse them. These functions basically do lexing, parsing of identifiers and string literals, things that are not overly specific to the Plush language.

One feature that is missing from the parsing library is the ability to parse floating-point numbers. I wrote a `parseInt` function, because that's relatively straightforward, but parsing numbers in general is trickier. The main reason why it's trickier is that all numbers start with a digit, and you don't know, when you begin parsing the number, if you are parsing a float or not.

The [current Plush implementation](https://github.com/zetavm/zetavm/blob/5f5a140de250292f93f54362442c750f3d5232b8/plush/plush_pkg.pls#L214) solves this problem by accumulating the digits in a string. This, however, is kind of bad, because it causes a series of successive string concatenations. I would like to have an implementation that avoids this.

IMO, the best strategy is probably to note the current index in the input string being parsed when starting to parse a number, and scan until either a period is hit, or the end of the digits are found. Then get the chars as a substring, and delegate the actual parsing to the $str_to_i32 or $str_to_f32 instructions. Possibly, integer parsing could be done directly, to completely avoid the allocation caused by the substring call.

There are other details to keep in consideration, such that some values might be too large to be parsed as valid integers. This should result in an exception being thrown (see `parseError` function in the parsing library).

Lastly, I would like, for now, to force floats to end with the `f` char. For example, `3.2f`. This is because we currently only support 32-bit floats, but may eventually support doubles.

**Comment by wjh** [comment]
Hello, I'm pretty interested in taking a look at this is that's ok?

**Comment by maximecb** [comment]
@wjh Certainly. Nobody has dared to try their hand at this issue yet. Go for it :)

**Comment by ivoyankulovski** [comment]
Try using BigInt -like implementation, resolving decimals into integer vector. Decimals are represented as direct int32 or higher. In the event of real float, decimal integer is divided by power of 10 or bit shifted according to number of digits and added to first elements of the vector.

**Comment by maximecb** [comment]
AFAIK, it's not that straightforward, there are pitfalls. I would rather delegate to the C implementation, but do it without allocating temporary strings/vectors.

**Comment by ivoyankulovski** [comment]
Exactly. Native configuration of the VM should allow selection of precision per language before compiling it, most end case scenarios include 3D rendering, currency computation in fintech software and prime number calculation in cryptography. IMO removing "f" at some point of time is important, while relying only on the decimal point existence, i.e. default float should be written as 0.0

**Comment by noundlas** [comment]
Is this feature implanted yet??

**Comment by maximecb** [comment]
@noundlas I don't believe it is

**Comment by Basile-z** [comment]
In a lexer I wrote I use two functions to add a token, which solve this problem:

1. `anticipateToken(TokenKind)` : this stores the position and the expected token kind.
2.   -  `validateToken(TokenKind)` : once done this allocates the token but correct the anticipated type, e.g after parsing an integer a dot was found, followed by more digits. Other example of use is to make an error token.
     -  `validateToken()`: once done with the token this overload allocates using the anticipation from 1.


([code](https://github.com/Basile-z/styx/blob/d0b452ce16560d754a8654495e56710c53f56fdc/src/styx/lexer.d#L43-L76))
